ext/php5/serializer.c

```diff
@@ -433,6 +441,113 @@ void ddtrace_serialize_span_to_array(ddtrace_span_fci *span_fci, zval *array TSR
+            break;
+        case E_USER_ERROR:
+            error_type = "E_USER_ERROR";
+            break;
+    }
+    MAKE_STD_ZVAL(*type);
+    ZVAL_STRING(*type, error_type, 1);
+}
+
+static void dd_fatal_error_msg(const char *format, va_list args, zval **msg) {
+    va_list args2;
+    char *buffer;
+
+    va_copy(args2, args);
+    int buffer_len = vspprintf(&buffer, 0, format, args2);
+    va_end(args2);
+
+    MAKE_STD_ZVAL(*msg);
+    if (buffer_len <= 0) {
+        ZVAL_STRING(*msg, "Unknown error", 1);
+        efree(buffer);
+        return;
+    }
+
+    /* In PHP 5 an uncaught exception results in a fatal error. The error
+     * message includes the call stack and its arguments, which is a vector for
+     * information disclosure. We need to avoid this.
+     */
+    const char uncaught[] = "Uncaught ";
+    /* The -1 is to avoid the null terminator which is included in literals and
+     * will not be present in the rendered error message at that position.
+     */
+    size_t uncaught_len = sizeof uncaught - 1;
+    if ((unsigned)buffer_len >= uncaught_len && memcmp(buffer, uncaught, uncaught_len) == 0) {
+        char *newline = memchr(buffer, '\n', buffer_len);
+        if (newline) {
+            *newline = '\0';
+            size_t linelen = newline - buffer;
+            // +1 for null terminator
+            ZVAL_STRINGL(*msg, buffer, linelen + 1, 1);
+        } else {
+            // This is suspect; there is always a newline in these messages
+            ZVAL_STRING(*msg, "Unknown uncaught exception", 1);
+        }
+    } else {
+        ZVAL_STRING(*msg, buffer, 1);
+    }
+    efree(buffer);
+}
+
+static void dd_fatal_error_stack(zval **stack) {
+    zval *trace;
+    TSRMLS_FETCH();
+
+    MAKE_STD_ZVAL(trace);
+    zend_fetch_debug_backtrace(trace, 0, DEBUG_BACKTRACE_IGNORE_ARGS, 0 TSRMLS_CC);
+    if (Z_TYPE_P(trace) == IS_ARRAY) {
+        char *stack_str = dd_serialize_stack_trace(trace TSRMLS_CC);
+        if (stack_str) {
+            MAKE_STD_ZVAL(*stack);
+            ZVAL_STRING(*stack, stack_str, 0);
+        }
+    }
+    zval_ptr_dtor(&trace);
+}
+
+typedef struct dd_error_info {
+    zval *type;
+    zval *msg;
+    zval *stack;
+} dd_error_info;
+
+static int dd_fatal_error_to_meta(zval *meta, dd_error_info error) {
+    if (error.type) {
+        Z_ADDREF_P(error.type);
+        add_assoc_zval(meta, "error.type", error.type);
+    }
+    if (error.msg) {
+        Z_ADDREF_P(error.msg);
+        add_assoc_zval(meta, "error.msg", error.msg);
+    }
+    if (error.stack) {
+        Z_ADDREF_P(error.stack);
+        add_assoc_zval(meta, "error.stack", error.stack);
+    }
+    return error.type && error.msg ? SUCCESS : FAILURE;
+}
+
+static dd_error_info dd_fatal_error(int type, const char *format, va_list args) {
+    dd_error_info error = {0};
+    dd_fatal_error_type(type, &error.type);
+    dd_fatal_error_msg(format, args, &error.msg);
+    dd_fatal_error_stack(&error.stack);
+    return error;
+}
```

Comment:
Is there a particular reason for passing the zval pointers as out parameters instead of using a return value? All 3 functions are void, so that would seem more natural, I think.

Reply:
Nope, no particular reason; I just got into the PHP 5 API mindset to out-param all the things. 😄 I agree that it is more natural to return. I'll update this.